test/test_block_adjacent_difference.cu

```diff
@@ -0,0 +1,521 @@
+
+    AssertEquals(expected_count, actual_count);
+  }
+}
+
+template <unsigned int ItemsPerThread,
+          unsigned int ThreadsInBlock>
+void TestBaseCase(unsigned int num_items,
+                  thrust::device_vector<CustomType> &d_data)
+{
+  thrust::tabulate(d_data.begin(), d_data.end(), IntToCustomType{1});
+
+  constexpr bool read_left  = true;
+  constexpr bool read_right = false;
+
+  BlockAdjacentDifferenceTest<CustomType,
+                              ItemsPerThread,
+                              ThreadsInBlock,
+                              read_left>(thrust::raw_pointer_cast(
+                                           d_data.data()),
+                                         num_items,
+                                         CustomType{});
+
+  {
+    const std::size_t expected_count = num_items;
+    const std::size_t actual_count =
+      thrust::count(d_data.begin(), d_data.end(), CustomType{1, 1});
+
+    AssertEquals(expected_count, actual_count);
+  }
+
+  thrust::tabulate(d_data.begin(), d_data.end(), IntToCustomType{});
+
+  BlockAdjacentDifferenceTest<CustomType,
+                              ItemsPerThread,
+                              ThreadsInBlock,
+                              read_right>(thrust::raw_pointer_cast(
+                                            d_data.data()),
+                                          num_items,
+                                          CustomType{num_items, num_items});
+
+  {
+    const auto unsigned_minus_one = static_cast<unsigned int>(-1);
+
+    const std::size_t expected_count = num_items;
+    const std::size_t actual_count =
+      thrust::count(d_data.begin(),
+                    d_data.end(),
+                    CustomType{unsigned_minus_one, unsigned_minus_one});
+
+    AssertEquals(expected_count, actual_count);
+  }
+}
+
+template <
+  typename ValueType,
+  unsigned int ItemsPerThread,
+  unsigned int ThreadsInBlock>
+void Test()
+{
+  constexpr int tile_size = ItemsPerThread * ThreadsInBlock;
+  thrust::device_vector<ValueType> d_values(tile_size);
+
+  for (unsigned int num_items = tile_size; num_items > 1; num_items /= 2)
+  {
+    d_values.resize(num_items);
+    TestLastTile<ValueType, ItemsPerThread, ThreadsInBlock>(num_items, d_values);
+  }
+
+  d_values.resize(tile_size);
+  Test<ValueType, ItemsPerThread, ThreadsInBlock>(tile_size, d_values);
+}
+
+template <
+  unsigned int ItemsPerThread,
+  unsigned int ThreadsInBlock>
+void TestCustomType()
+{
+  constexpr int tile_size = ItemsPerThread * ThreadsInBlock;
+  thrust::device_vector<CustomType> d_values(tile_size);
+
+  for (unsigned int num_items = tile_size; num_items > 1; num_items /= 2)
+  {
+    d_values.resize(num_items);
+    TestLastTile<ItemsPerThread, ThreadsInBlock>(num_items, d_values);
+  }
+
+  d_values.resize(tile_size);
+  TestBaseCase<ItemsPerThread, ThreadsInBlock>(tile_size, d_values);
+}
+
+template <unsigned int ItemsPerThread, unsigned int ThreadsPerBlock>
+void Test()
```

Comment:
Reminder to test in-place `BlockAdjacentDifference` after updating the implementation/docs to support this.